nixos/modules/security/gnupg.nix

```diff
@@ -0,0 +1,253 @@
+    description = "Socket for gpg-agent";
+    wantedBy = ["sockets.target"];
+    socketConfig.ListenStream = "${gnupgHome}/S.gpg-agent";
+    socketConfig.SocketMode = "0600";
+  };
+  environment.systemPackages =
+    # TODO: maybe this would be better to do that directly in pkgs.gnupg?
+    let gpgPresetPassphrase = pkgs.runCommand "gpg-preset-passphrase"
+      { preferLocalBuild = true;
+        allowSubstitutes = false;
+      } ''
+      mkdir -p $out/bin
+      ln -s -t $out/bin ${pkgs.gnupg}/libexec/gpg-preset-passphrase
+    ''; in
+    [ pkgs.gnupg gpgPresetPassphrase ];
+  systemd.services =
+    lib.optionalAttrs cfg.enableGpgAgent {
+      gpg-agent = {
+        description = "gpg-agent for decrypting GnuPG-protected secrets";
+        requires = ["gpg-agent.socket"];
+        serviceConfig = {
+          Type = "simple";
+          # Because /run/user/0 is wiped out by pam_systemd when root logouts,
+          # systemd.services.gpg-agent cannot put its socket in
+          # the path expected by gpg:
+          # /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts
+          # derived from ${gnupgHome}.
+          #
+          # But GPG_AGENT_INFO is ignored with gpg >= 2.1,
+          # hence this hack makes gpg connect to ${gnupgHome}/S.gpg-agent
+          # by relaxing permissions of the expected socket directory.
+          # With this hack, uploading the passphrase should now be doable with just:
+          # ssh root@example.org 'gpg-preset-passphrase --homedir /var/lib/gnupg --preset $keygrip'
+          #
+          # An other way would be to remove /run/user/0 with:
+          # BindPaths = ["/var/empty:/run/user"];
+          # but this would also require a more complex passphrase uploading:
+          # ssh root@example.org "unshare --mount sh -c '
+          #   mount --bind /var/empty /run/user;
+          #   gpg-preset-passphrase --homedir /var/lib/gnupg --preset $keygrip'"
+          ExecStartPre = "${pkgs.coreutils}/bin/install -D -d -m 640 /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts";
+          ExecStart = ''${pkgs.gnupg}/bin/gpg-agent \
+           --supervised \
+           --homedir '${gnupgHome}' \
+           --allow-loopback-pinentry \
+           --allow-preset-passphrase \
+           ${lib.escapeShellArgs cfg.gpgAgentFlags}
+          '';
+          Restart = "on-failure";
+          RestartSec = 5;
+          StateDirectory = ["gnupg"];
+          StateDirectoryMode = "700";
+        };
+      };
+    } //
+    lib.mapAttrs' (target: secret:
+      lib.nameValuePair (lib.removeSuffix ".service" secret.service) {
+        description = "Install secret ${secret.path}";
+        after = ["gpg-agent.service"];
+        wants = ["gpg-agent.service"];
+        script = ''
+          set -o pipefail
+          set -eux
+          decrypt() {
+            # In case /run/user/0 has been cleaned up by pam_systemd
+            install -D -m 640 -d /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts
```

Comment:
Again this seems racy. What uses case do you want to support beyond having gpg simply working?

Reply:
This is here for the `gpg --decrypt` just after, in the case where the `install` in `ExecStartPre=` in `gpg-agent.service` has been wiped out by a root login/logout but root is logged in again so `/run/usr/0` is here again. The only race I see is between this install and the `gpg` right after, which seems highly improbable.
I thought it could also help for the uploading `gpg`/`gpg-preset-passphrase` commands but as you pointed out there was a problem, and now I require this `install` also before the uploading commands.

Reply:
Is the `BindPaths` above an appropriate replacement?

Reply:
Hmm, strangely enough my comment about removing completely `/run/user/0` is wrong now, it can just be empty, but I could have sworn emptying `/run/user/0` was not enough in my tests 12h ago. *shrugs*
```
$ ssh root@localhost
# gpg -K
# file /run/user/0/gnupg/S.gpg-agent
/run/user/0/gnupg/S.gpg-agent: socket
```
Both `unshare` and `system-run` work fine.
```
# unshare --mount sh -c "mkdir -p /run/user/0/gnupg && mount --bind /var/empty /run/user/0/gnupg && gpgconf --homedir $PWD --list-dirs agent-socket"
/var/lib/gnupg/S.gpg-agent
```
```
# systemd-run -P -p BindPaths='/var/empty:/run/user/0/gnupg' gpgconf --homedir /var/lib/gnupg --list-dirs agent-socket
Running as unit: run-u4801.service
/var/lib/gnupg/S.gpg-agent
```
I'll change the `install` hack to that `BindPaths=` hack, making it `BindReadOnlyPaths=`, in the services and in the uploading through `systemd-run -P`.

Reply:
Ok. The other alternative would be to get a proper logind session with machinectl as noted above.

Reply:
Hmm, this works also: `InaccessiblePaths=/run/user/0` but not: `InaccessiblePaths=/run/user/0/gnupg`. Strange.

Reply:
Oh, ok, `InaccessiblePaths=/run/user/0` does work either  if `/run/user/0` does not exist. So AFAICS that could only be `InaccessiblePaths=/run/user` or `BindReadOnlyPaths=/var/empty:/run/user/0/gnupg`

Reply:
Good news! In fact `pam_systemd` uses a unit to setup `/run/user/0` and an `ExecStartPost=` running the `install` hack can be added to the instance of the template used, by writing to `/etc/systemd/system/user-runtime-dir@0.service.d/override.conf` (which requires another hack in NixOS, but is doable). Also, because `ExecStartPost=` is taken into account for the purpose of `Before=`/`After=` ordering constraints, this prevents race condition regarding the uploading commands. And for the decrypting services I've kept the `BindReadOnlyPaths=` hack.